SimPEG/Mesh/CylMesh.py

```diff
@@ -383,46 +428,67 @@ def _fastInnerProductDeriv(self, projType, prop, invProp=False, invMat=False):
             :rtype: function
             :return: dMdmu, the derivative of the inner product matrix
         """
-        assert projType in ['F', 'E'], "projType must be 'F' for faces or 'E' for edges"
+        assert projType in ['F', 'E'], ("projType must be 'F' for faces or 'E'"
+                                        " for edges")
         tensorType = Utils.TensorType(self, prop)
 
         dMdprop = None
 
-        if invMat:
-            MI = self._fastInnerProduct(projType, prop, invProp=invProp, invMat=invMat)
+        if invMat or invProp:
+            MI = self._fastInnerProduct(projType, prop, invProp=invProp,
+                                        invMat=invMat)
 
         if tensorType == 0:
             Av = getattr(self, 'ave'+projType+'2CC')
             V = Utils.sdiag(self.vol)
-            ones = sp.csr_matrix((np.ones(self.nC), (range(self.nC), np.zeros(self.nC))), shape=(self.nC,1))
+            ones = sp.csr_matrix((np.ones(self.nC), (range(self.nC),
+                                                     np.zeros(self.nC))),
+                                 shape=(self.nC,1))
+            nelts = np.sum(getattr(self, 'vn'+projType).nonzero())
+
             if not invMat and not invProp:
-                dMdprop = self.dim * Av.T * V * ones
+                dMdprop = nelts * Av.T * V * ones
             elif invMat and invProp:
-                dMdprop =  self.dim * Utils.sdiag(MI.diagonal()**2) * Av.T * V * ones * Utils.sdiag(1./prop**2)
+                dMdprop = nelts * (Utils.sdiag(MI.diagonal()**2) * Av.T * V *
+                                   ones * Utils.sdiag(1./prop**2))
+            elif invProp:
+                dMdprop = nelts * Av.T * V * Utils.sdiag(- 1./prop**2)
+            elif invMat:
+                dMdprop = nelts * Utils.sdiag(- MI.diagonal()**2) * Av.T * V
 
         if tensorType == 1:
             Av = getattr(self, 'ave'+projType+'2CC')
             V = Utils.sdiag(self.vol)
+            nelts = np.sum(getattr(self, 'vn'+projType).nonzero())
+
             if not invMat and not invProp:
-                dMdprop = self.dim * Av.T * V
+                dMdprop = nelts * Av.T * V
```

Comment:
is `nelts` (number of elements, 1 for edges, 2 for faces on cyl mesh) a clear enough variable name? any other suggestions